sim/pokemon.ts

```diff
@@ -19,29 +19,80 @@ interface MoveSlot {
 }
 
 export class Pokemon {
-	side: Side;
-	battle: Battle;
-	set: PokemonSet;
-	getHealth: (side: Side) => string;
-	getDetails: (side: Side) => string;
+	readonly side: Side;
+	readonly battle: Battle;
+
+	/** "pre-bound" functions for nicer syntax allows them to be passed directly to Battle#add. */
+	readonly getHealth: (side: Side) => string;
+	readonly getDetails: (side: Side) => string;
+
+	readonly set: PokemonSet;
+	readonly name: string;
+	readonly fullname: string;
+	readonly id: string; // shouldn't really be used anywhere
+	readonly species: string;
+	readonly speciesid: string;
+	readonly level: number;
+	readonly gender: GenderName;
+	readonly happiness: number;
+	readonly pokeball: string;
+
+	/** Transform copies IVs in gen 4 and earlier, so we track the base IVs/HP-type/power. */
+	readonly baseIvs: StatsTable;
+	readonly baseHpType: string;
+	readonly baseHpPower: number;
+
+	readonly baseMoveSlots: MoveSlot[];
+	moveSlots: MoveSlot[];
+
+	hpType: string;
+	hpPower: number;
+
+	position: number;
+	details: string;
+
 	baseTemplate: Template;
-	species: string;
-	name: string;
-	speciesid: string;
 	template: Template;
-	moveSlots: MoveSlot[];
-	baseMoveSlots: MoveSlot[];
+	speciesData: AnyObject;
+
+	status: string;
+	statusData: AnyObject;
+	volatiles: AnyObject;
+	showCure: boolean;
+
 	baseStats: StatsTable;
+	stats: {[k: string]: number};
```

Comment:
Why aren't these a `StatsTable`?

Reply:
```suggestion
	/**
	 * These are pre-modification stored stats in-battle. At switch-in,
	 * they're identical to `baseStats`, but can be temporarily changed
	 * until switch-out by effects such as Power Trick and Transform.
	 *
	 * Stat multipliers from abilities, items, and volatiles, such as
	 * Solar Power, Choice Band, or Swords Dance, are not stored in
	 * `stats`, but applied on top and accessed by `pokemon.getStat`.
	 *
	 * `stats` are reset to `baseStats` whenever volatiles are cleared
	 * (i.e. on switch-out).
	 */
	stats: StatsTable;
```

Reply:
I would probably rename these to `storedStats` and `baseStoredStats`; to make it clearer that they aren't the Pokémon's "actual" stats.

Reply:
You haven't answered the other question, though – why isn't this a `StatsTable`?

Reply:
This type was what it original was in the JSdoc, ands its not `StatsTable` because theres no `hp` (see line 322).

We could change it to `SparseStatsTable` which seems a little better (but results in *beaucoup* type errors to hunt down), but maybe we should instead have `hp` in there and just add a `get maxhp() { return this.stats.hp; }` method?

Reply:
I'd add `get maxhp() { throw ... }` myself, but either works.

Reply:
How would people get the max HP then? https://github.com/Zarel/Pokemon-Showdown/blob/master/sim/pokemon.ts#L424

Reply:
Oh, I meant in the `stats` table.

```js
this.stats = {
  get maxhp() { throw ... },
  atk: 0,
  ...
}
```

Reply:
`calculateStat` should probably also `throw` instead of forwarding to `this.maxhp`.

Or actually we should use the client approach, `StatNameExceptHP`.